# Petrolera OilyGiant

Se nos ha solicitado encontrar los mejores lugares para abrir 200 pozos nuevos de petróleo.
Para la tarea nos ha proporcionado datos sobre la calidad del crudo y el volumen de reservas de cada pozo petrolero de tres regiones.

El objetivo principal es crear un modelo que permita elegir la región con el mayor beneficio.

Tenemos 3 dataset uno por cada región, como primer paso preprocesaremos la información (valores nan, duplicados, tipo de datos). 

Posterior a ello crearemos un modelo para predecir el volumen de reservas de pozos nuevos, con base en los resultados seleccionaremos los pozos que tienen mayor volumen de reserva estimado, con esta selección definiremos la región con el mayor beneficio.

 ## Importación de librerías, descarga y preprocesamiento de los datos

### Importación y lectura de datos

In [ ]:
# importar librerias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [ ]:
# leer datos
geo_0 = pd.read_csv('./geo_data_0.csv')
geo_1 = pd.read_csv('./geo_data_1.csv')
geo_2 = pd.read_csv('./geo_data_2.csv')

In [5]:
geo_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [6]:
geo_1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [7]:
geo_2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [8]:
geo_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [9]:
geo_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [10]:
geo_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [11]:
geo_0.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

In [12]:
geo_1.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

In [13]:
geo_2.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

In [14]:
geo_0.duplicated().sum()

0

In [15]:
geo_1.duplicated().sum()

0

In [16]:
geo_2.duplicated().sum()

0

Los 3 dataset no tienen valores ausentes, tampoco duplicados y el formato de columnas y tipo de datos es el correcto.

Las columnas `f0, f1, f2` que corresponden a características de la calidad del crudo varian de una región a otra, sin embargo dado que es la misma característica en cualquier región se entiende que es una métrica estandarizada, razón por la cual no vamos a estandarizar dichas columnas.

Vamos a pasar la creación del modelo.

## Modelización de datos

Debido a que tenemos 3 dataset con las mismas columnas, cada una para una región diferente, vamos a crear una función que automatice los procesos para predecir el volumen de reservas y la RMSE.

In [ ]:
def modela(df):
    # divide las características y el objetivo
    features = df.drop(['product', 'id'], axis=1)
    target = df['product']

    # segmenta el 25% de los datos para hacer el conjunto de validación
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25,
                                                                                  random_state=12345)
    # crea el modelo de regresión lineal
    model = LinearRegression()
    model.fit(features_train, target_train)

    # predice deacuerdo al modelo ajustado
    predictions = model.predict(features)

    return pd.Series(target), pd.Series(predictions)

In [18]:
def imprimir(target, predictions):
    # calcula el volumen medio de reservas predicho y la RMSE en el conjunto de validación
    rmse = mean_squared_error(target, predictions) ** 0.5
    print("RMSE:", rmse, ", Volumen medio de reservas predichas", predictions.mean())

In [ ]:
target_0, predict_0 = modela(geo_0)
imprimir(target_0, predict_0)

RMSE: 37.69240596744535 , Volumen medio de reservas predichas 92.62849276088778


In [ ]:
target_1, predict_1 = modela(geo_1)
imprimir(target_1, predict_1)

RMSE: 0.8903802650297566 , Volumen medio de reservas predichas 68.82635271777153


In [ ]:
target_2, predict_2 = modela(geo_2)
imprimir(target_2, predict_2)

RMSE: 40.055624818752136 , Volumen medio de reservas predichas 95.02020328978764


Para la región geográfica 0 el volumen medio de reservas predichas es: `92.62849276088778`. Con un RMSE de: `37.69240596744535`.

Para la región geográfica 1 el volumen medio de reservas predichas es: `68.82635271777151`. Con un RMSE de: `0.890380265029757`.

Para la región geográfica 2 el volumen medio de reservas predichas es: `95.02020328978764`. Con un RMSE de: `40.055624818752136`.

La región 0 y 2 tienen un RMSE mucho mayor que la región 1, lo que significa que hay una diferencia considerable entre las predicciones y el valor real. Caso que la región 1 no tiene tan marcado.

## Calculo de ganancias

Pasemos a la métrica de investigación, la ganancia.

Para este proyecto la petrolera va a invertir 100 millones de dolares para el desarrollo de 200 nuevos pozos, lo que quiere decir que la inversión promedio para cada pozo es de 500 mil dolares. Si no queremos perdidas el ingreso promedio para cada pozo debería ser de mínimo 500 mil dolares para que el beneficio o perdida fuera de 0.

Pero ¿qué tendría que suceder para que un pozo generara 500 mil dolares?.
El volumen de reservas representa la cantidad de barriles (miles) de cada pozo, cada barril genera 4.5 dolares, esto da 4500 dolares por cada un mil barriles.
Entonces para generar 500 mil dolares con cada mil barriles de 4500 dolares, el promedio de los pozos debería tener 111.1 miles de barriles de petróleo.

En síntesis para generar un margen de ganancia positivo el promedio de volumen de reservas debe ser mayor a 111.1 miles de barriles.

In [ ]:
inversion = 100000000
no_pozos = 200
costo_unidad = 4500

# inversión promedio por cada pozo
inv_pozo = inversion/no_pozos

# volumen de reservas promedio minimmo para generar 500 mil dolares.
vol_pozo = inv_pozo / costo_unidad

In [23]:
vol_pozo

111.11111111111111

Tomando como referencia para que haya beneficios positivos el volumen medio de reservas debe ser mayor a 111.1.

Si comparamos este referente con la media del volumen predicho de las 3 regiones, ninguna media de las regiones lo supera, por lo que vamos a necesitar otra estrategia mejor que la media para lograr el mejor beneficio posible.

De cada región vamos a seleccionar los 200 pozos con los valores de predicción más altos, de los mejores pronósticos tomamos sus valores y lo multiplicamos por el costo de cada unidad de volumen de reservas y obtenemos el ingreso, a ese resultado le descontamos los costos, que son 100 millones. Si de la selección de los 200 pozos de cada región, su media fue mayor a 111.1 el beneficio va ser positivo.

In [ ]:
def beneficio(target, predictions, count):
    pred_sorted = predictions.sort_values(ascending=False)
    selected = target.reindex(index=pred_sorted.index[:count])

    # beneficio en millones de dolares
    return (selected.sum() * costo_unidad) - inversion

In [25]:
beneficio(target_0, predict_0, no_pozos)

34941041.92112014

In [26]:
beneficio(target_1, predict_1, no_pozos)

24150866.966815114

In [27]:
beneficio(target_2, predict_2, no_pozos)

25714106.319609284

Evaluando el beneficio en un escenario hipotético donde nos guiaramos por los pozos con la mayor reserva disponible, eligiendo sus valores reales, la región 0 sería la apta para desarrollar los 200 pozos pretroleros ya que tuvo un beneficio de casi 35 millones, más beneficio que el generado por la región 1 y 2.

Esta evaluación de beneficio es en el caso hipótetico donde nos guiaramos por los pronósticos más altos de toda la región, sin embargo esto no sucede en la realidad.

Vamos a simular 1000 selecciones aleatorias donde vamos a elegir a los mejores 200 pozos de una submuestra de 500 pozos, calculamos el beneficio de esas 1000 simulaciones, obtenemos el beneficio media con un intervalo de confianza al 95 % y el % de pérdida.

Vamos a analizar los resultados de las 3 regiones y elegiremos las regiones que tengan una perdida debajo de 2.5% y beneficio medio más alto.

In [ ]:
def bootstrap(target, predict):
    state = np.random.RandomState(12345)
    n = 1000
    values = []
    size_sample = 500
    count = 0

    for i in range(n):
        pred_sample = predict.sample(
            n=size_sample, replace=True, random_state=state)
        targ_sample = target.reindex(index=pred_sample.index)
        pred_sample.reset_index(inplace=True, drop=True)
        targ_sample.reset_index(inplace=True, drop=True)
        benef = beneficio(targ_sample, pred_sample, no_pozos)
        values.append(benef)
        if benef < 0:
            count += 1

    # distribucion de los beneficios
    values = pd.Series(values)

    # beneficio promedio
    mean = values.mean()

    # intervalo de confianza del 95%
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)

    # riesgo de perdidas
    loss = 100 * (count / n)

    print("Ganancia promedio:", mean)
    print("Intervalo de confianza 95%:", lower, '-', upper)
    print("Riesgo de pérdidas:", loss, '%')

In [29]:
 bootstrap(target_0, predict_0)

Ganancia promedio: 4241497.51404505
Intervalo de confianza 95%: -778569.2373488933 - 9685806.995986093
Riesgo de pérdidas: 5.2 %


In [30]:
 bootstrap(target_1, predict_1)

Ganancia promedio: 4487206.400344507
Intervalo de confianza 95%: 374980.7453396168 - 8238299.202083874
Riesgo de pérdidas: 1.4000000000000001 %


In [31]:
 bootstrap(target_2, predict_2)

Ganancia promedio: 3739390.2198985177
Intervalo de confianza 95%: -1819481.884471201 - 8962994.081008721
Riesgo de pérdidas: 9.2 %


Realizando una simulación para evaluar lo que puede pasar en un plano real, la región que se ajustó a las condiciones dadas es la región 1 cuyo riesgo de pérdida es de 1.4% y coincidentemente un beneficio mayor que las otras dos regiones de`4487206.4` con un intervalo de confianza del 95%.

La selección de la región óptima en esta fase (región 1) no coincidió con el elegido considerando todos los pozos de la región donde se seleccionó la región 0.

## Conclusiones

La selección de la región donde desarrollar los 200 pozos guiándonos por el cálculo de las ganancias considerando todos los pozos de la región (región 0 elegida) no coincidió con las ganancias promedio usando el método boostrap (región 1 elegida).

Hacer uso de la técnica bootstrap nos ayudó a crear la distribución de las ganancias y con ello obtener la ganancia esperada en una situación real, a si solo nos basaramos en las mejores ganancias de cada región, porque pueden haber puntos atípicos que hagan que la selección se distorcione.

En conclusión para la Petrolera OilyGiant se sugiere desarrollar los 200 pozos petroleros en la región 1 con un beneficio promedio de + 4 millones y un riesgo de pérdidas de 1.4%.